File: Lab5/Task2/CSharp/MyString.cs
using System;

namespace CSharp
{
    public class MyString
    {
        protected char[] Value;
        public string GetValue() => string.Join("", Value);

        public MyString(char[] str)
        {
            this.Value = str;
        }

        protected internal virtual int GetLength()
        {
            return this.Value.Length;
        }

        public virtual void Insert(char elem, int index)
        {    
            var newValue = new char[GetLength() + 1];
            Array.Copy(this.Value, 0, newValue, 0, GetLength());
            for (var i = index; i < Value.Length; i++)
            {
                newValue[i + 1] = Value[i];
            }

            newValue[index] = elem;
            this.Value = newValue;
        }
    }
}

File: Lab5/Task2/C++/C++/MyString.cpp
#include "MyString.hpp"
#include <cstring>

using namespace std;

MyString::MyString(char* str){
    this->Value = str;
}

string MyString::GetValue(){
    return this->Value;
}

int MyString::GetLength() {

    return strlen(this->Value);
}

void MyString::Insert(char elem, uint index) {
    int length = this->GetLength();
    char* newValue = new char[length + 1];
    for (int i = 0; i < length + 1; i++) {
        if ( i < index ) {
            newValue[i] = Value[i];
        }
        if ( i == index ) {
            newValue[i] = elem;
        }
        if ( i > index ){
            newValue[i] = Value[i-1];
        }
    }
    Value = newValue;
}

